storage/ndb/rest-server2/server/src/rdrs_rondb_connection_pool.cpp

```diff
@@ -91,16 +129,61 @@ RS_Status RDRSRonDBConnectionPool::AddMetaConnections(
   return RS_OK;
 }
 
-RS_Status RDRSRonDBConnectionPool::GetNdbObject(Ndb **ndb_object) {
+RS_Status RDRSRonDBConnectionPool::GetNdbObject(Ndb **ndb_object,
+                                                Uint32 threadIndex) {
   RS_Status status = Check();
   if (unlikely(status.http_code != SUCCESS)) {
     return status;
   }
-  return dataConnection->GetNdbObject(ndb_object);
+  ThreadContext *thread_context = m_thread_context[threadIndex];
+  require(threadIndex < m_num_threads);
+  NdbMutex_Lock(thread_context->m_thread_context_mutex);
+  if (likely(thread_context->m_is_shutdown == false &&
+             thread_context->m_is_ndb_object_in_use == false &&
+             thread_context->m_ndb_object != nullptr)) {
+    /**
+     * This is by far the most common path through the code. We are not
+     * shutting down, the Ndb object isn't in use and it exists. We only
+     * allocate the Ndb object normally on the first interaction.
+     */
+    *ndb_object = thread_context->m_ndb_object;
+    thread_context->m_is_ndb_object_in_use = true;
+    NdbMutex_Unlock(m_thread_context[threadIndex]->m_thread_context_mutex);
+    return RS_OK;
+  }
+  require(thread_context->m_is_ndb_object_in_use == false);
```

Comment:
If this is correct then why not move it to the top of the function?

Reply:
One could do so, yep